### ISSUE TITLE
-1 as a default missing value does not work when the actual annotation value could be -1

### ISSUE THREAD

**Issue by dvg-p4** [brentp/echtvar#28]
Specifically, I'm working with CADD scores, which as a signed beta value (or something like that), can be any real number (though realistically seems to range from -3 to +30). I notice that echtvar is including variants which weren't found in the archive in the output with "-1" as the value; I'm guessing this is the code that does that:

https://github.com/brentp/echtvar/blob/b53778957b28ae6461d5bce0d3904818316d4019/src/lib/fields.rs#L30-L32

Since -1 is a possible actual value for this field, it does not suffice as an indicator of missingness. (For now, I can probably work around this since the scaled CADD scores *are* nonnegative, but if I was annotating from a source where all columns could possibly be -1, I'd be out of luck.) 

Glancing at [the BCF standard](https://samtools.github.io/hts-specs/VCFv4.2.pdf), it looks like the canonical way to encode a missing value is as a "signaling NaN" (`0x7F800001`) for floats and as the minimum possible value (`0x80000000`) (-2^31) for integers.

**Comment by dvg-p4** [comment]
...oddly though, the json file for generating the cadd archive specifies "missing = -2"

**Comment by dvg-p4** [comment]
though github doesn't seem to think this is valid json

**Comment by brentp** [comment]
Hi, echtvar uses hjson, so it allows comments and not quoting field names. Indeed, you can choose your own missing value by using the JSON. 
If you indicate that indeed you have specified, for example INT32_MIN as the missing value and instead echtvar saves -1, then I will look into this. It tries to do something sane for e.g. allele frequency so we can use AF < 0.001 and missing vlaues will also pass.

**Comment by dvg-p4** [comment]
Thanks for the quick response! I think the issue might be, the json is only passed to `echtvar encode` not `echtvar anno`? The input CADD tsv's don't have any missing values, but the vcf I'm annotating has variants that aren't in CADD's tsv. Let me experiment with changing the json params.

**Comment by dvg-p4** [comment]
Oh I see the encoded zip stores a config json within itself...but it doesn't actually seem to match!
`my_cadd.json`:
```hjson
[{
        "field": "raw",
        "alias": "cadd_raw",
        multiplier: 100,
        zigzag: true,
        missing: -2147483648,
}, {
        field: "phred",
        alias: "cadd_phred",
        multiplier: 100,
        zigzag: false,
        missing: -2147483648,
}]
```
```bash
$ echtvar encode test_cadd.indel.v1.6.hg38.zip my_cadd.json cadd.indel.v1.6.bcf
[...] [succeeds]
$ unzip test_cadd.indel.v1.6.hg38.zip
[inflating everything to ./echtvar]
$ cat echtvar/config.json
[
  {
    "field": "raw",
    "alias": "cadd_raw",
    "missing_value": -1,
    "missing_string": "MISSING",
    "zigzag": true,
    "multiplier": 100,
    "ftype": "Float"
  },
  {
    "field": "phred",
    "alias": "cadd_phred",
    "missing_value": -1,
    "missing_string": "MISSING",
    "zigzag": false,
    "multiplier": 100,
    "ftype": "Float"
  }
]
```

**Comment by dvg-p4** [comment]
oh hmm that's missing_**value** and missing_**string**, not just `missing`. Let's try matching that in the input json...
```
$ cat my_cadd_2.json
[{
        "field": "raw",
        "alias": "cadd_raw",
        multiplier: 100,
        zigzag: true,
        missing_value: -2147483648,
}, {
        field: "phred",
        alias: "cadd_phred",
        multiplier: 100,
        zigzag: false,
        missing_value: -2147483648,
}]
$ echtvar encode test_2_cadd.indel.v1.6.hg38.zip my_cadd_2.json cadd.indel.v1.6.bcf
$ unzip test_2_cadd.indel.v1.6.hg38.zip
$ cat echtvar/config.json
[
  {
    "field": "raw",
    "alias": "cadd_raw",
    "missing_value": -2147483648,
    "missing_string": "MISSING",
    "zigzag": true,
    "multiplier": 100,
    "ftype": "Float"
  },
  {
    "field": "phred",
    "alias": "cadd_phred",
    "missing_value": -2147483648,
    "missing_string": "MISSING",
    "zigzag": false,
    "multiplier": 100,
    "ftype": "Float"
  }
]
```
aha!

**Comment by brentp** [comment]
yeah, thanks for noting this. Not sure why you didn't get a parsing error. I will update the cadd readme

**Comment by dvg-p4** [comment]
```
##bcftools_viewCommand=view PVCF_470K_real_CADD_indel_test_2.bcf; Date=Fri May 26 14:50:03 2023
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO
1       69026   1:69026:T:G     T       G       .       .       cadd_raw=-2.14748e+09;cadd_phred=-2.14748e+09
1       69081   1:69081:G:C     G       C       .       .       cadd_raw=-2.14748e+09;cadd_phred=-2.14748e+09
1       69095   1:69095:T:G     T       G       .       .       cadd_raw=-2.14748e+09;cadd_phred=-2.14748e+09
1       69134   1:69134:A:G     A       G       .       .       cadd_raw=-2.14748e+09;cadd_phred=-2.14748e+09
```
And looks like that value is now actually carried over. It would be great if there were a way to actually have it set to NaN (and come out as such, or as `.` when converting to human-readable format) but this is definitely good enough to work with. And I noticed glancing through the repo that you're doing some hacky stuff storing floats as integers and such, which I presume is one of the reasons why I'm nitpicking your output format while CADD.sh continues to chug for hours ;-p In any case, adding on a `-i PHRED>=0` will probably end up doing what I actually want, which is omitting those rows. Thanks!

**Comment by brentp** [comment]
It seems the json5 lib that I am using allows extra fields not defined in the struct. So that makes it difficult to error-check this case. :(

I haven't done it, but it may be possible to get a true missing value. htslib uses this:
uint32_t bcf_float_missing    = 0x7F800001;

and then this: https://github.com/samtools/htslib/blob/a59bc5256ee4f539ae47edd575dc1340356d922a/htslib/vcf.h#L1467-L1472

The hacky stuff is to allow storing floats as strings. But it still should be quite fast. It's just that with CADD, you have around 9 billion variants and so reading+writing also takes a while.

**Comment by dvg-p4** [comment]
All I'm saying is, whatever it is you're doing runs fast haha. (At least comparatively so--7 hours to convert to bcf, then about 2 hours to build the encoding, and only a minute or two to annotate. Meanwhile, CADD's convoluted snakemake pipeline is at 20 hours and counting for one annotation job.

I think to get "proper" missing values for floats, you'd have to have a special case when converting the intermediate representation back to float, and hardcode in that hex constant for those cases. Dunno how feasible that is in Rust, it's not a language I'm familiar with.

**Comment by brentp** [comment]
I think it doesn't actually mess with the missing value, even for floats (most values get mutliplied and truncated) so it might work as-is.
Glad it's fast anway! And that it's working for your needs. Let me know any other problems.

**Comment by brentp** [comment]
closed by #32
